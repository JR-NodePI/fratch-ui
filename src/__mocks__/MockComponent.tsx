const renderPropValue = (value: any) => {
  switch (typeof value) {
    case 'function':
      return 'function () {}';
    case 'object':
      return JSON.stringify(value);
    default:
      return value;
  }
};

export default function MockComponent({
  __testDisplayName,
  children,
  ...restProps
}: any): JSX.Element {
  return (
    <>
      <code aria-label={`${__testDisplayName ?? ''} props`}>
        {Object.entries(restProps).map(
          ([key, value]) => `${key} : ${renderPropValue(value)}`
        )}
      </code>
      {children && (
        <code aria-label={`${__testDisplayName ?? ''} children`}>
          {children}
        </code>
      )}
    </>
  );
}
