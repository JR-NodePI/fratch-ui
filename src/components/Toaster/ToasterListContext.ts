import React from 'react';

export interface Toaster {
  id?: string;
  title?: string;
  message: string;
  type: 'success' | 'error' | 'warning' | 'info';
  duration?: number;
  nlToBr?: boolean;
}

export interface ToasterListProviderProps {
  addToaster?: (toaster: Toaster) => void;
  toasters?: Toaster[];
}

const ToasterListContext = React.createContext<ToasterListProviderProps>({});
ToasterListContext.displayName = 'ToasterListContext';

export default ToasterListContext;
