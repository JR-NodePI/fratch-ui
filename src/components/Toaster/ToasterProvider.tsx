import { type ReactNode, useCallback, useRef, useState } from 'react';
import { createPortal } from 'react-dom';

import { v4 as uuid } from 'uuid';

import { c } from '../../helpers/classNameHelpers';
import ToasterItem from './ToasterItem';
import ToasterListContext, { type Toaster } from './ToasterListContext';

import styles from './Toaster.module.css';

export default function ToasterProvider({
  children,
  listClassName,
  itemClassName,
}: {
  children: ReactNode;
  listClassName?: string;
  itemClassName?: string;
}): JSX.Element {
  const [portalId] = useState<string>(uuid());
  const [toasters, setToasters] = useState<Toaster[]>([]);

  const handleToasterClose = (id: string): void => {
    setToasters(toasters.filter(toaster => toaster.id !== id));
  };

  const addToaster = useCallback((toaster: Toaster): void => {
    setToasters((prevToasters: Toaster[]) => [
      ...prevToasters,
      { ...toaster, id: uuid() },
    ]);
  }, []);

  const toasterListRef = useRef<HTMLDivElement>(null);

  return (
    <ToasterListContext.Provider value={{ toasters, addToaster }}>
      {createPortal(
        toasters.length > 0 && (
          <div
            ref={toasterListRef}
            className={c(styles.toaster_list, listClassName)}
          >
            {toasters.map(toaster => (
              <ToasterItem
                className={c(itemClassName)}
                key={toaster.id}
                onClose={handleToasterClose}
                toaster={toaster}
              />
            ))}
          </div>
        ),
        document.body,
        portalId
      )}
      {children}
    </ToasterListContext.Provider>
  );
}
