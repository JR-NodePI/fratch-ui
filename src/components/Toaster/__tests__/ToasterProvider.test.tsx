import { useContext } from 'react';

import { act, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import MockComponent from '../../../__mocks__/MockComponent';
import ToasterListContext from '../ToasterListContext';
import { Toaster, ToasterItemProps } from '../ToasterProps';
import ToasterProvider from '../ToasterProvider';

let triggerCloseToaster: (toasterId: string) => void;
vi.mock('../ToasterItem', () => ({
  default: ({ onClose, ...props }: ToasterItemProps): JSX.Element => {
    triggerCloseToaster = onClose;
    return <MockComponent __testDisplayName="ToasterItem" {...props} />;
  },
}));

describe('ToasterProvider', () => {
  const onClose = vi.fn();
  const initialToaster: Toaster = { type: 'info', duration: 100 };
  const initialProps = {
    onClose,
    toaster: initialToaster,
    itemClassName: 'mock-item-class',
  };
  let triggerAddToaster: (toasters: Toaster) => void;

  const MockContent = (): JSX.Element => {
    const { toasters, addToaster } = useContext(ToasterListContext);
    triggerAddToaster = addToaster;
    return <p>{`Toaster counters: ${toasters?.length}`}</p>;
  };

  const setup = (
    props: ToasterItemProps = initialProps
  ): ReturnType<typeof render> =>
    render(
      <ToasterProvider {...props}>
        <MockContent />
      </ToasterProvider>
    );

  let randomUUIDCounter = 0;
  beforeEach(() => {
    vi.spyOn(crypto, 'randomUUID').mockImplementation(() => {
      randomUUIDCounter++;
      return `crypto-mock-random-UUID-${randomUUIDCounter}`;
    });
  });

  afterEach(() => {
    randomUUIDCounter = 0;
    vi.restoreAllMocks();
    onClose.mockClear();
  });

  it('should render properly', () => {
    const { baseElement } = setup();
    expect(baseElement).toMatchSnapshot();
  });

  it('should add toasters', () => {
    const { baseElement } = setup();

    act(() => {
      triggerAddToaster({ ...initialToaster, type: 'info' });
      triggerAddToaster({ ...initialToaster, type: 'success' });
      triggerAddToaster({ ...initialToaster, type: 'error' });
      triggerAddToaster({ ...initialToaster, type: 'warning' });
    });

    expect(baseElement).toMatchSnapshot();
  });

  it('should close a toaster', () => {
    const { baseElement } = setup();

    act(() => {
      triggerAddToaster({ ...initialToaster, type: 'info' });
    });

    expect(baseElement).toMatchSnapshot('with initial toaster');

    act(() => {
      triggerCloseToaster('crypto-mock-random-UUID-2');
    });

    expect(baseElement).toMatchSnapshot('with deleted toaster');
  });
});
