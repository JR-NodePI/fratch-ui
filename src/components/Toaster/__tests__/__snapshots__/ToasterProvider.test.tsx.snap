// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToasterProvider > should add toasters 1`] = `
<body>
  <div>
    <p>
      Toaster counters: 4
    </p>
  </div>
  <div
    class="_toaster_list_68dafd"
  >
    <code
      aria-label="ToasterItem props"
    >
      className : mock-item-class
      toaster : {"type":"info","duration":100,"id":"crypto-mock-random-UUID-2"}
    </code>
    <code
      aria-label="ToasterItem props"
    >
      className : mock-item-class
      toaster : {"type":"success","duration":100,"id":"crypto-mock-random-UUID-4"}
    </code>
    <code
      aria-label="ToasterItem props"
    >
      className : mock-item-class
      toaster : {"type":"error","duration":100,"id":"crypto-mock-random-UUID-5"}
    </code>
    <code
      aria-label="ToasterItem props"
    >
      className : mock-item-class
      toaster : {"type":"warning","duration":100,"id":"crypto-mock-random-UUID-6"}
    </code>
  </div>
</body>
`;

exports[`ToasterProvider > should close a toaster > with deleted toaster 1`] = `
<body>
  <div>
    <p>
      Toaster counters: 0
    </p>
  </div>
</body>
`;

exports[`ToasterProvider > should close a toaster > with initial toaster 1`] = `
<body>
  <div>
    <p>
      Toaster counters: 1
    </p>
  </div>
  <div
    class="_toaster_list_68dafd"
  >
    <code
      aria-label="ToasterItem props"
    >
      className : mock-item-class
      toaster : {"type":"info","duration":100,"id":"crypto-mock-random-UUID-2"}
    </code>
  </div>
</body>
`;

exports[`ToasterProvider > should render properly 1`] = `
<body>
  <div>
    <p>
      Toaster counters: 0
    </p>
  </div>
</body>
`;
